### PR TITLE
added ISO15693 (NfcV) reading, saving, emulating and revealing from privacy mode (unlock)

### PR DESCRIPTION
**Pull request by g3gg0**
# What's new

This PR is related to #1991 (without all its history), built on top of #2070 (candidate for merge) and from a new branch.

- Added NfcV through various layers
- Added NfcV memory and system info reading
- Added loading/saving/emulating
- Added unlocking NfcV type NXP SLIX-L (reveal from privacy mode)
- Usable for another module already in development by @nvx (picopass emulation)

This is a PR, being aware that:
- rfalNfcv* calls are being made from lib/ level
- Nfc code is undergoing a refactoring

Q: Might someone of the maintainers help touching some of the quesionable code sections, as i am still not very confident with
the inner workings etc?

I consider the code being good enough for a PR process, few reworking will have to be done. I you disagree, please tell.

regards,
g3gg0

### CODE REVIEW

**Comment by nemanjan00** [comment]
READ_MULTI_BLOCKS seems to be broken:

```
Flipper:

        384 |      59264 | Rdr |62  23  ff  ff  ff  ff  00  00  07  e0  00  05  aa  eb                   |  ok | READ_MULTI_BLOCK(0-5)
      63296 |     177984 | Tag |00  12  34  56  78  90  12  34  56  78  90  12  34  56  78  90  12  34   |     | 
            |            |     |56  78  90  12  34  56  78  04  84                                       |  ok | 

Real card:
        384 |      59264 | Rdr |62  23  ff  ff  ff  ff  00  00  07  e0  00  05  aa  eb                   |  ok | READ_MULTI_BLOCK(0-5)
      63552 |     202816 | Tag |00  00  12  34  56  78  00  90  12  34  56  00  78  90  12  34  00  56   |     | 
            |            |     |78  90  12  00  34  56  78  90  00  12  34  56  78  55  31               |  ok | 
```

**Comment by g3gg0** [comment]
thanks, didn't notice and will have a look at it.

**Comment by g3gg0** [comment]
this is in unadressed mode:
grafik

here in adressed mode:
grafik

here with 1oo256 mode:
grafik

why is the custom bit set in your request?

**Comment by nemanjan00** [comment]
20230120_00h53m48s_grim

Looks like you are missing block security status

**Comment by g3gg0** [comment]
from which datasheet are you citing?

**Comment by nemanjan00** [comment]
BS_ISO_IEC_15693-3-2009__A2-2015.pdf

**Comment by g3gg0** [comment]
ah okay i have a quite old standard revision.
seems the block security status is requested when the custom flag is set (?)

need to find a latest datasheet. if someone knows where to find it -> git - at - g3gg0.de

**Comment by nemanjan00** [comment]
Guessing you will probably write it better, but, anyway...

```patch

diff --git a/lib/nfc/protocols/nfcv.c b/lib/nfc/protocols/nfcv.c
index cc7f4ad5..2ffdd0ea 100644
--- a/lib/nfc/protocols/nfcv.c
+++ b/lib/nfc/protocols/nfcv.c
@@ -537,6 +537,29 @@ void nfcv_emu_handle_packet(
             nfcv_emu_send(
                 tx_rx, nfcv_data, ctx->response_buffer, 1, ctx->response_flags, ctx->send_time);
         } else {
+            if(ctx->command == ISO15693_READ_MULTI_BLOCK) {
+                ctx->response_buffer[0] = ISO15693_NOERROR;
+
+                for(int current_block = 0; current_block < blocks; current_block++) {
+                    memcpy(
+                        &ctx->response_buffer[1 + (1 + nfcv_data->block_size) * current_block + 1],
+                        &nfcv_data->data[nfcv_data->block_size * (block + current_block)],
+                        nfcv_data->block_size);
+
+                        ctx->response_buffer[1 + (1 + nfcv_data->block_size) * current_block] = 0;
+                }
+
+                nfcv_emu_send(
+                    tx_rx,
+                    nfcv_data,
+                    ctx->response_buffer,
+                    1 + (nfcv_data->block_size + 1) * blocks,
+                    ctx->response_flags,
+                    ctx->send_time);
+
+                break;
+            }
+
             ctx->response_buffer[0] = ISO15693_NOERROR;
             memcpy(
                 &ctx->response_buffer[1],
```

**Comment by g3gg0** [comment]
thanks, when trying a real SLIX-L it looks like that:
grafik

so if the custom bit is set, it seems the card will send the security status.
can you check the standard if this bit 6 (0x40) in the flags is set, the security status is prepended?

**Comment by nemanjan00** [comment]
Yes, according to spec, Block security status is optional and depends on `Option_flag`

20230120_01h11m15s_grim

Also I have sent you email with spec pdf

**Comment by nemanjan00** [comment]
Also, I assume TwoSubcarrier is not supported?

**Comment by g3gg0** [comment]
that is correct. this is not supported (yet).
see nfcv_emu_alloc() line 284
grafik
here i allocate the waveforms for high and low data rate.
subcarrier modulation would have to get added here.

**Comment by nemanjan00** [comment]
```
Filetype: Flipper NFC device
Version: 3
# Nfc device type can be UID, Mifare Ultralight, Mifare Classic or ISO15693
Device type: ISO15693
# UID is common for all formats
UID: E0 07 00 00 FF FF FF FF
# Data Storage Format Identifier
DSFID: 00
# Application Family Identifier
AFI: 00
IC Reference: 01
# Number of memory blocks, usually 0 to 256
Block Count: 28
# Size of a single memory block, usually 4
Block Size: 04
Data Content: 12 34 56 78 90 12 34 56 78 90 12 34 56 78 90 12 34 56 78 90 12 34 56 78 90 12 34 56 78 90 12 34 56 78 90 12 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
# Subtype of this card (0 = ISO15693, 1 = SLIX, 2 = SLIX-S, 3 = SLIX-L, 4 = SLIX2)
Subtype: 00
# End of ISO15693 parameters
```

Just in case you need sample I used that made PM3 use that flag

**Comment by g3gg0** [comment]
first/third real SLIX-L, second/fourth flipper
grafik

**Comment by nemanjan00** [comment]
I use:
```
hf 15 rdmulti -* -b 0 --cnt 6
```

**Comment by g3gg0** [comment]
fixed that part, thanks a lot!

**Comment by nemanjan00** [comment]
Thank you!

**Comment by nemanjan00** [comment]
One of my readers causes Flipper to crash will NULL pointer when I open Log. I will be sniffing traffic tomorrow and sending more info

**Comment by g3gg0** [comment]
thanks. in the evening i will check where NULL dereference might have happened.

**Comment by g3gg0** [comment]
@nemanjan00 
I pushed a few changes. maybe you can check if you still see crashes?
thanks!

**Comment by g3gg0** [comment]
> One of my readers causes Flipper to crash will NULL pointer when I open Log. I will be sniffing traffic tomorrow and sending more info

this issue is also fixed. that was the case if the only received commands were addresses and the log was opened.

**Comment by g3gg0** [comment]
ok, added a few more optional commands, just to make it more complete.
not sure which systems use them, but now DSFID and AFI should be handled as the standard suggests.

**Comment by gornekich** [comment]
Hello @g3gg0 !
Thanks for your work on NFC-V support! Since this PR is based on #2070 , I convert this PR to draft until your previous PR is merged. However ,I am starting to test and review this PR now

**Comment by nvx** [comment]
Rebased onto latest version of #2070 (which in turn was rebased onto the latest dev branch). Once #2070 is merged this should be good to go as well.

**Comment by g3gg0** [comment]
@nvx  you tell me when to issue this PR for review again?

**Comment by nvx** [comment]
> @nvx you tell me when to issue this PR for review again?

Will do. I figure worth leaving it as draft until #2070 is merged, then a quick rebase if necessary and can submit for review then.

**Comment by g3gg0** [comment]
for those who have issues with the ISO15693 emulation, there is also some issue when reading with mobile phones using e.g. "NFC Tools".
however from the looks @nvx and me are quite certain this feature sends the correct data, but the phone cannot detect the bits.
could as well be the antenna or the NFC chips modulation rate causing issues.

**Comment by nvx** [comment]
Branch updated to latest dev and api symbols updated (both f7 and f18 targets). Functionality tested fine. Ready for merge or feedback :)

**Comment by nvx** [comment]
Branch updated to latest dev and api symbols updated (both f7 and f18 targets). Functionality tested fine. Ready for merge or feedback :)

**Comment by g3gg0** [comment]
here as reference a comparison of flipper' response (top) and an original SLIX-L (bottom) to an IDENTIFY request.

you see the reader's request last bits on the left side and the load modulation on the right side
grafik

zoomed in a bit, shows the modulation which initially looked very promising, but zooming into it, flipper's modulation is a bit more noisy. 
might come from the general RF noise emitted by flipper, not sure. thats nothing that can (reasonably) be influenced by these patches.
grafik

edit:
to prevent misunderstandings: i don't think that causes any issues.
it's just the signal is not 100% as clean as those from tags.

**Comment by g3gg0** [comment]
when checking proxmark's emulation, they don't seem to have the modulation synchronized to the main field.
the last partial dip in field strength makes me believe so.
which means the modulation we generate looks comparable.

grafik

**Comment by g3gg0** [comment]
made some updates which would probably would have come up during your review anyway @gornekich 
we are now using NFCV_ prefixes and also not using RFAL_ defines for non-rfal calls.
also changed all(?) constants to definitions

using "inverse modulation" compared to a real tag like e.g. proxmark does increase detection range.

**Comment by nvx** [comment]
Is there anything blocking this from being merged that g3gg0 or myself need to address?

**Comment by skotopes** [comment]
@nvx Planning to review it today

File: applications/main/nfc/scenes/nfc_scene_nfc_data_info.c
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_nfc_data_info.c]: I suggest using nfc_util_bytes2num() instead of this function

File: lib/nfc/protocols/nfcv.h
Comment (gornekich) [inline lib/nfc/protocols/nfcv.h]: Could you explain where 0.9998f comes from?

**Comment by g3gg0** [comment]
@skotopes hope no comment is a good sign and you are happy with the code as it is? :D

**Comment by skotopes** [comment]
@g3gg0 @gornekich is reviewing this PR, also please merge dev into this branch

File: lib/nfc/nfc_worker.c
Comment (gornekich) [inline lib/nfc/nfc_worker.c]: Please rework with FURI_LOG_D() everywhere

File: applications/main/nfc/scenes/nfc_scene_nfcv_emulate.c
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_nfcv_emulate.c]: delete please =)

File: applications/main/nfc/scenes/nfc_scene_read.c
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_read.c]: I would add NfcSceneNfcvReadSuccess. There I would display NFC-V type, UID Block number and Block size. Everything should fit on the screen without scrolling. Also add "More" button there, and in NfcSceneNfcvMenu add "Info", which will switch to NfcSceneNfcDataInfo scene

File: applications/main/nfc/scenes/nfc_scene_nfc_data_info.c
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_nfc_data_info.c]: If you add NfcSceneNfcvReadSuccess scene which I mentioned above, there shouldn't be "More" button on Info scene. If I load Nfcv card from Saved menu, there is already Emulate option in SavedMenu scene.

**Review by gornekich — CHANGES_REQUESTED** [review]
Great work, thanks!
Could you please make some fixes and merge dev

**Review by gornekich — CHANGES_REQUESTED** [review]
Great work, thanks!
Could you please make some fixes and merge dev

**Comment by gornekich** [comment]
Also in #2529 we introduce new NFC architecture. MfUltralight is almost reworked with new API, now we are working on MfClassic and MfDesfire. I think we will finish low level API in a couple of weeks and continue to add support of other protocols.

However API is not finalized yet, it would be nice to receive some comments from you. And if you are interested in reworking NFC-V support with new API in future, please let me know =)

File: lib/nfc/protocols/nfcv.h
Comment (g3gg0) [reply lib/nfc/protocols/nfcv.h]: Should be removed, yeah.
Was for compensating the frequencies being off a bit.
Could be systematic, could be xtal tolerances.

**Comment by nvx** [comment]
Will update with the requested changes tonight. Thanks for the review!

**Comment by nvx** [comment]
Merged dev and made all the suggested fixes except the GUI changes. I'm not super familiar with the Flipper GUI stuff so that'll be a little longer for me to get that part done unless someone felt like beating me to it.

**Comment by xMasterX** [comment]
> Merged dev and made all the suggested fixes except the GUI changes. I'm not super familiar with the Flipper GUI stuff so that'll be a little longer for me to get that part done unless someone felt like beating me to it.

Hi @nvx 

I made all requested GUI changes, and fixed crash that occurs when you exit from emulation and start it again
I don't see how I can make PR into PR, so I'm going to share two patches, please take them

[b0c322.patch](https://github.com/flipperdevices/flipperzero-firmware/files/11636215/b0c322.patch)
[2dc005.patch](https://github.com/flipperdevices/flipperzero-firmware/files/11636216/2dc005.patch)

**Comment by nvx** [comment]
@xMasterX awesome! I'll get those patches applied shortly!

**Comment by nvx** [comment]
Patches merged and tested. Also merged in latest dev again as well. Cheers for the GUI patches and nice catch with that crash fix @xMasterX too!

I believe this closes out all the outstanding items from the review. Let me know if anything else is spotted.

**Comment by g3gg0** [comment]
thanks @nvx @xMasterX for your great support here! :)

also @nvx for hunting down that "hmmm android doesn't read flipper that reliable anymore?" issue, which was a pain!

**Review by gornekich — DISMISSED** [review]

**Review by skotopes — APPROVED** [review]

**Comment by skotopes** [comment]
@g3gg0 that was some serious journey, my congratulations.

**Comment by team-orangeBlue** [comment]
and it's merged. cool!

**Comment by g3gg0** [comment]
wow, great!!
thanks a lot to you and all contributors!

**Comment by eychei** [comment]
Hi everyone,

@g3gg0 
I am trying to emulate a Slix2 tag with a different originality signature.

This is from the datasheet:

• Originality signature:
32 byte ECC based originality signature

It is set by NXP in production and can be changed for customers. 

The Dymo label writers do have a different signature thus I can not emulate the tag.
This is a link to the forum post: https://www.eevblog.com/forum/reviews/dymo-550-thermal-printer-drm-hacking/

This is a dump of a tag with the originality signature (the signature is different for every UID it seems):

inventory: 0x01,0x5C,0x63,0x7B,0x3F,0x08,0x01,0x04,0xE0
sysInfo: 0x0F,0x5C,0x63,0x7B,0x3F,0x08,0x01,0x04,0xE0,0x01,0x3D,0x4F,0x03,0x01
signature: 0x6A,0x67,0x2D,0xAE,0x8D,0xFB,0xC7,0xC2,0x12,0x65,0x49,0x41,0x9B,0x3C,0x5F,0xAF,0xD7,0xC6,0xC0,0x97,0x33,0x43,0x99,0xE5,0xD0,0x5A,0xD9,0x71,0xFE,0x7B,0x74,0xA9
blocks:
0x03,0x0A,0x82,0xED,0x86,0x39,0x61,0xD2,0x03,0x14,0x1E,0x32,0xB6,0xCA,0x00,0x3C,0x62,0x64,0xCE,0x0F,0x33,0x30,0x32,0x35,0x36,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xFF,0x24,0x01,0x01,0x00,0x00,0x00,0x22,0x04,0x1E,0x00,0x23,0x00,0x00,0x00,0x00,0x00,0x10,0x00,0xF8,0x03,0x4B,0x02,0x00,0x00,0x00,0x00,0x6B,0x02,0x2C,0x01, 0x04,0x3E,0x1E,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xD7,0xFA,0x00,0x1C,0x63,0xF2,0x1B,0xE0,0x04,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xDD,0x4B,0x9D,0x05,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x32,0x8C,0x00,0x30,0x06,0xE2,0xDA,0x89, 0x00,0x00,0x00,0x00,0x6B,0x22,0xD6,0x76,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x11,0xF3,0x00,0x2C,0xDD,0xC3,0x3E,0x91,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xB6,0xFE,0x00,0x01

Hope you can help out.

-e

**Comment by g3gg0** [comment]
a) as I didnt check that stuff yet, I cannot tell anything.
b) i can add those extended SLIX2 features (statically) and supply some test releases
c) send me a tag and I can check if it behaves as the original.

Please note that the signature seems to be a ECC signature and emulating it (i.e. generating matching it to the emulated UID) is nearly impossible.
Asymmetric ciphers are designed to not be cracked by hobbyists btw :)

NXP AN11600 describes how to verify the signature, but it is restricted.
**If anyone has it, feel free to mail  iso15693<at>g3gg0.de**

**Comment by eychei** [comment]
Thanks.
It seems that if we just emulate the signature (memory) nothing needs to be computed.
The reader firmware will check if the signature is correct. 
Where can I send you a tag to? Will do asap.

-e

**Comment by g3gg0** [comment]
ah, found [this](https://github.com/free-dmo/free-dmo-stm32/blob/main/Src/main.c).

yes, that can be emulated easily i guess.

**Comment by eychei** [comment]
I think one little bit is also missing.

The read password needs to be set to increment the counter. It is set for each tag differently. I can sniff the password for one of my tags and we could test on that data.

-e

**Comment by g3gg0** [comment]
if the code accepts any password, wouldn't that work as well?

**Comment by g3gg0** [comment]
the [code says](https://github.com/free-dmo/free-dmo-stm32/blob/main/Src/main.c#L492) so

**Comment by eychei** [comment]
The firmware needs the correct password. It is generated for each UID. 
I do not have the generation scheme. 
The firmware is locked and I can not read it out. The STM32F072 is in RDP Level 2.
I do have a chipwhisperer but can not get it down to Level 1 or 0.

It may be that the counter will not work but the firmware may ignore that.
We could have the generate random number command send out the same number every time and the set password command always return a 1 (OK).

OH!
You are right. This is not needed. Just return the signal as true (success).

**Comment by g3gg0** [comment]
can you start a feature request / issue for this?

**Comment by eychei** [comment]
Yes will do now. 

-e

**Comment by eychei** [comment]
Added feature request here:

https://github.com/flipperdevices/flipperzero-firmware/issues/2781#issue-1761400779

**Comment by semaf** [comment]
@g3gg0 

What is the point of "Can not save key file" after I try to save the read data (tonies)? Emulating is working. It still save it but get the error "Can not parse file".

Content of the .nfc file:
```
Filetype: Flipper NFC device
Version: 3
Nfc device type can be UID, Mifare Ultralight, Mifare Classic, FeliCa or ISO15693
Device type: ISO15693
```

I managed to read before with your fork that saved .nfc and .shd files but now the official firmware nor the rogue firmware is working. Not tested with your latest fork on github. Emulating on official firmware is not working but on Rogue firmware.

**Comment by skotopes** [comment]
@semaf do you have proxmark ? can you send us card dump ?

**Comment by g3gg0** [comment]
Unfortunately I don't have my flipper at the moment.
Back then, the code did unlock and save the. nfc file immediately.
Seems something has changed and the saving code doesn't work anymore. Thus you get the error I guess?

**Comment by semaf** [comment]
> @semaf do you have proxmark ? can you send us card dump ?

Yes I do but will have time the next days..

**Comment by semaf** [comment]
> Unfortunately I don't have my flipper at the moment.
> 
> Back then, the code did unlock and save the. nfc file immediately.
> 
> Seems something has changed and the saving code doesn't work anymore. Thus you get the error I guess? 

The latest unleashed firmware version is working without any issue. Saving .nfc files.

**Comment by bitboxer** [comment]
Is there a reason this is no longer on the dev branch of Flipper? This PR https://github.com/flipperdevices/flipperzero-firmware/pull/3050 removed it all?

**Comment by team-orangeBlue** [comment]
Well, yes.
The NFC refactor blew up all of NFC. It now works for itself and nothing else.
That means, the app will work, but
* `nfc apdu` is gone
* Outside implementations, like nested/NFC magic are gone
* other firmwares (lol) just didn't merge this and when one did (MH-HM) the dev said his compiler exploded

So uhhh... yeah. But we still have ISO15693!

**Comment by gornekich** [comment]
@bitboxer We are working on this functionality now. Will be ready soon

**Comment by g3gg0** [comment]
didn't follow the development since merge.
do you need support from my side?

**Comment by gornekich** [comment]
@g3gg0 thanks, I think I will do it by my self and once it's done I will ask you to test it

**Pull request merged**
